File: Depth_First_Search/DFS_slidesMarcos.java
package Depth_First_Search;

import java.util.ArrayList;

public class DFS_slidesMarcos {
    static final int BRANCO = -1;
    static final int CINZA = 0;
    static final int PRETO = 1;

    static ArrayList<Integer> cor;
    static ArrayList<Integer> i;
    static ArrayList<Integer> f;
    static ArrayList<Integer> ante;
    static int tempo = 1;

    static void DFS_VISIT(ArrayList<ArrayList<Integer>> grafo, int u) {
        cor.set(u, CINZA);
        i.set(u, tempo++);

        for (int v : grafo.get(u)) {
            if (cor.get(v) == BRANCO) {
                ante.set(v, u);
                DFS_VISIT(grafo, v);
            }
        }

        cor.set(u, PRETO);
        f.set(u, tempo++);
    }

    static void DFS_START(ArrayList<ArrayList<Integer>> grafo, int s) {
        cor = new ArrayList<>(grafo.size());
        i = new ArrayList<>(grafo.size());
        f = new ArrayList<>(grafo.size());
        ante = new ArrayList<>(grafo.size());
        tempo = 1;

        for (int j = 0; j < grafo.size(); j++) {
            cor.add(BRANCO);
            i.add(-2);
            f.add(-2);
            ante.add(-2);
        }

        DFS_VISIT(grafo, s);
    }

    public static void main(String[] args) {
        // Exemplo de grafo (lista de adjacência)
        ArrayList<ArrayList<Integer>> grafo = new ArrayList<>();
        grafo.add(new ArrayList<>() {{ add(1); add(4); }});
        grafo.add(new ArrayList<>() {{ add(0); add(2); add(4); }});
        grafo.add(new ArrayList<>() {{ add(1); add(3); }});
        grafo.add(new ArrayList<>() {{ add(2); add(4); add(5); }});
        grafo.add(new ArrayList<>() {{ add(0); add(1); add(3); }});
        grafo.add(new ArrayList<>() {{ add(3); }});

        // Executando o algoritmo DFS começando a partir do vértice 0
        DFS_START(grafo, 0);

        // Exibindo os resultados
        System.out.println("Resultado do algoritmo DFS:");
        System.out.println("Vertice:   Cor:    Inicio:    Fim:    Antecessor:");
        for (int v = 0; v < grafo.size(); ++v) {
            System.out.println(v + "          " + cor.get(v) + "       " + i.get(v) + "         " + f.get(v) + "        " + ante.get(v));
        }
    }
}

File: Depth_First_Search/DFS_slidesMarcos.cpp
#include <iostream>
#include <vector>

using namespace std;

const int BRANCO = -1;
const int CINZA = 0;
const int PRETO = 1;

vector<int> cor;
vector<int> i;
vector<int> f;
vector<int> ante;
int tempo = 1;

void DFS_VISIT(vector<vector<int>>& grafo, int u) {
    cor[u] = CINZA;
    i[u] = tempo++;

    for (int v : grafo[u]) {
        if (cor[v] == BRANCO) {
            ante[v] = u;
            DFS_VISIT(grafo, v);
        }
    }

    cor[u] = PRETO;
    f[u] = tempo++;
}

void DFS_START(vector<vector<int>>& grafo, int s) {
    cor.assign(grafo.size(), BRANCO);
    i.assign(grafo.size(), -2);
    f.assign(grafo.size(), -2);
    ante.assign(grafo.size(), -2);
    tempo = 1;

    DFS_VISIT(grafo, s);
}

int main() {
    // Exemplo de grafo (lista de adjacência)
    vector<vector<int>> grafo = {
        {1, 4},     // vértice 0
        {0, 2, 4},  // vértice 1
        {1, 3},     // vértice 2
        {2, 4, 5},  // vértice 3
        {0, 1, 3},  // vértice 4
        {3}         // vértice 5
    };

    // Executando o algoritmo DFS começando a partir do vértice 0
    DFS_START(grafo, 0);

    // Exibindo os resultados
    cout << "Resultado do algoritmo DFS:" << endl;
    cout << "Vertice:   Cor:    Inicio:    Fim:    Antecessor:" << endl;
    for (int v = 0; v < grafo.size(); ++v) {
        cout << v << "          " << cor[v] << "       " << i[v] << "         " << f[v] << "        " << ante[v] << endl;
    }

    return 0;
}